File: app/src/main/java/com/example/movies/ui/detail/DetailActivity.java
package com.example.movies.ui.detail;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.example.movies.R;
import com.example.movies.core.base.BaseActivity;
import com.example.movies.core.util.Constants;
import com.example.movies.data.entity.Backdrops;
import com.example.movies.data.entity.DetailInfo;
import com.example.movies.data.entity.GalleryPhotoModel;
import com.example.movies.data.entity.Result;
import com.example.movies.data.service.RetroFitService;

import java.util.ArrayList;

import butterknife.BindView;
import de.hdodenhof.circleimageview.CircleImageView;
import me.relex.circleindicator.CircleIndicator;


public class DetailActivity extends BaseActivity {

    @BindView(R.id.info_tab_layout)
    TabLayout tabLayout;
    @BindView(R.id.view_pager)
    ViewPager viewPager;
    @BindView(R.id.circularIndicator)
    CircleIndicator circularIndicator;
    @BindView(R.id.profile_image)
    CircleImageView profileImage;
    @BindView(R.id.release_date)
    TextView releaseDate;
    @BindView(R.id.movie_name)
    TextView movieName;
    @BindView(R.id.view_pager_info)
    ViewPager viewPagerInfo;

    private int id;

    RetroFitService service = new RetroFitService();

    @Override
    public Integer getLayoutId() {
        return R.layout.activity_detail;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Intent i = getIntent();
        id = i.getIntExtra("id", 0);

        setViewPager();
        fetchMovieDetail();
        fetchGetBackDrops();
    }

    void fetchMovieDetail() {
        RetroFitService.ResultCallBack serviceCallBack = new RetroFitService.ResultCallBack() {
            @Override
            public void getResult(Result result) {
                Result<DetailInfo> detailInfoResult = result;
                if (detailInfoResult.getData() != null) {
                    setData(detailInfoResult.getData());
                } else {
                    Toast.makeText(DetailActivity.this, result.getErrorMessage(), Toast.LENGTH_LONG).show();
                }
            }
        };
        service.fetchMovieDetail(id, serviceCallBack);
    }

    void fetchGetBackDrops() {
        RetroFitService.ResultCallBack serviceCallBack = new RetroFitService.ResultCallBack() {
            @Override
            public void getResult(Result result) {
                Result<GalleryPhotoModel> galleryPhotoModelResult = result;

                if (galleryPhotoModelResult.getData() != null) {
                    setAdapter(galleryPhotoModelResult.getData().getBackdrops());
                } else {
                    Toast.makeText(DetailActivity.this, result.getErrorMessage(), Toast.LENGTH_LONG).show();
                }
            }
        };
        service.fetchBackDrops(id, serviceCallBack);
    }

    private void setData(DetailInfo list) {
        releaseDate.setText(list.getRelease_date());
        movieName.setText(list.getTitle());
        String title = list.getTitle();
        getSupportActionBar().setTitle(title);
        Glide.with(this).load(Constants.IMAGE_BASE_PATH + list.getPoster_path()).centerCrop().into(profileImage);

    }

    private void setAdapter(ArrayList<Backdrops> list) {
        PagerAdapter pagerAdapter = new PagerAdapterSlide(this, list);
        viewPager.setAdapter(pagerAdapter);
        circularIndicator.setViewPager(viewPager);
    }

    private void setViewPager() {
        DetailTabAdapter detailTabAdapter = new DetailTabAdapter(getSupportFragmentManager(), this, id);
        viewPagerInfo.setAdapter(detailTabAdapter);
        tabLayout.setupWithViewPager(viewPagerInfo);
    }
}

File: app/src/main/java/com/example/movies/ui/detail/DetailActivity.kt
package com.example.movies.ui.detail

import android.os.Bundle
import android.widget.Toast
import butterknife.ButterKnife
import com.bumptech.glide.Glide
import com.example.movies.R
import com.example.movies.core.base.BaseActivity
import com.example.movies.core.util.Constants
import com.example.movies.data.entity.Backdrops
import com.example.movies.data.entity.DetailInfo
import com.example.movies.data.entity.GalleryPhotoModel
import com.example.movies.data.entity.Result
import com.example.movies.data.service.RetroFitService
import kotlinx.android.synthetic.main.activity_detail.*
import java.util.*


class DetailActivity : BaseActivity() {

    private var id: Int = 0

    internal var service = RetroFitService()

    override val layoutId: Int?
        get() = R.layout.activity_detail

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ButterKnife.bind(this)
        val i = intent
        id = i.getIntExtra("id", 0)

        setViewPager()
        fetchMovieDetail()
        fetchGetBackDrops()
    }

    internal fun fetchMovieDetail() {
        val serviceCallBack = object : RetroFitService.ResultCallBack {
            override fun getResult(result: Result<*>) {
                val detailInfoResult: Result<DetailInfo> = result as Result<DetailInfo>
                if (detailInfoResult.data != null) {
                    setData(detailInfoResult.data!!)
                } else {
                    Toast.makeText(this@DetailActivity, detailInfoResult.errorMessage, Toast.LENGTH_LONG).show()
                }
            }
        }
        service.fetchMovieDetail(id, serviceCallBack)
    }

    internal fun fetchGetBackDrops() {
        val serviceCallBack = object : RetroFitService.ResultCallBack {
            override fun getResult(result: Result<*>) {
                val galleryPhotoModel: Result<GalleryPhotoModel> = result as Result<GalleryPhotoModel>
                if (galleryPhotoModel.data != null) {
                    setAdapter(galleryPhotoModel.data!!.backdrops)
                } else {
                    Toast.makeText(this@DetailActivity, galleryPhotoModel.errorMessage, Toast.LENGTH_LONG).show()
                }
            }
        }
        service.fetchBackDrops(id, serviceCallBack)
    }

    private fun setData(list: DetailInfo) {
        releaseDate.setText(list.release_date)
        movieName.setText(list.title)
        val title = list.title
        supportActionBar!!.setTitle(title)
        Glide.with(this).load(Constants.IMAGE_BASE_PATH + list.poster_path!!).centerCrop().into(profileImage)

    }

    private fun setAdapter(list: ArrayList<Backdrops>?) {
        val pagerAdapter = PagerAdapterSlide(this, list)
        viewPager.setAdapter(pagerAdapter)
        circularIndicator.setViewPager(viewPager)
    }

    private fun setViewPager() {
        val detailTabAdapter = DetailTabAdapter(supportFragmentManager, this, id)
        viewPagerInfo.setAdapter(detailTabAdapter)
        tabLayout.setupWithViewPager(viewPagerInfo)
    }
}
